### PR TITLE
Cuts-from-proofs: refactor the interface

### PR DESCRIPTION
**Pull request by aehyvari**
Move sparse matrix representation to a separate file, and remove
dependency on ArithLogic.

This version still uses LVRef as the index type for Polynomial inside the matrix operations.

### CODE REVIEW

File: src/tsolvers/lasolver/SparseMatrix.cc [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    opensmt::Real sum = 0;
+    for (auto const & term : poly) {
+        uint32_t index = term.var.x;
+        assert(index < values.size());
+        sum += term.coeff * values[index];
+    }
+    return sum;
+}
+
+SparseColMatrix::TermVec SparseColMatrix::Col::toVector() const {
```
Comment (blishko) [inline src/tsolvers/lasolver/SparseMatrix.cc:29]: Not sure we need this `TermVec`, it's basically a restricted version of `Polynomial`.
I am a bit worried about the unnecessary copying, but it's probably not a big deal...

File: src/tsolvers/lasolver/CutCreator.cc [hunk 2]
```diff
@@ -274,11 +26,11 @@ PTRef CutCreator::cut(std::vector<DefiningConstaint> const & constraints) {
     // an integer, the row represents an infeasible constraint
     for (uint32_t rowIndex = 0; rowIndex < dim; ++rowIndex) {
         auto const & row = matrixU[rowIndex];
-        auto product = crossProduct(row, varValues);
+        auto product = row.product(varValues);
         if (not product.isInteger()) {
-            return infeasibleRowToCut(row, columnMapping, logic, product);
+            return std::pair{row.toVector(), product};
```
Comment (blishko) [inline src/tsolvers/lasolver/CutCreator.cc:31]: Maybe we could use `opensmt::pair`? :)

File: src/tsolvers/lasolver/LASolver.cc [hunk 3]
```diff
@@ -855,11 +855,105 @@ bool LASolver::shouldTryCutFromProof() const {
...
+
+struct DefiningConstraint {
+    PTRef lhs;
+    opensmt::Real rhs;
+};
+
+/*
+ * Translates the set of defining constraints (linear equalities) into a suitable inner representation:
+ * - Matrix A of coefficients
+ * - a vector with right-hand-side values (as Rationals)
```
Comment (blishko) [inline src/tsolvers/lasolver/LASolver.cc:868]: Maybe replace these two points with a combined one mentioning the `SparseLinearSystem`?

File: src/tsolvers/lasolver/LASolver.cc [hunk 4]
```diff
@@ -855,11 +855,105 @@ bool LASolver::shouldTryCutFromProof() const {
...
+        LVRef var = term.first;
+        auto const & coeff = term.second;
+        args.push(logic.mkTimes(toVarMap[var.x], logic.mkIntConst(coeff)));
+    }
+    return logic.mkPlus(std::move(args));
+}
+
+PTRef cutToSplit(CutCreator::Cut && cut, std::vector<PTRef> const & toVarMap, ArithLogic & logic) {
+    auto termVec = cut.first;
+    auto rhs = cut.second;
```
Comment (blishko) [inline src/tsolvers/lasolver/LASolver.cc:940]:
```suggestion
    auto const & termVec = cut.first;
    auto const & rhs = cut.second;
```
To avoid copies.

**Review by blishko — COMMENTED** [review]
I think this is good! Small suggestions to avoid copies, but otherwise it should be good to go.
You should also add yourself to the license headers now :)

File: src/tsolvers/lasolver/CutCreator.cc (hunk 2 above)
Comment (aehyvari) [reply src/tsolvers/lasolver/CutCreator.cc:31]: Ok, absolutely!

File: src/tsolvers/lasolver/LASolver.cc (hunk 3 above)
Comment (aehyvari) [reply src/tsolvers/lasolver/LASolver.cc:868]: I'll try something..

File: src/tsolvers/lasolver/LASolver.cc (hunk 4 above)
Comment (aehyvari) [reply src/tsolvers/lasolver/LASolver.cc:940]: Ok.

File: src/tsolvers/lasolver/SparseMatrix.cc (hunk 1 above)
Comment (aehyvari) [reply src/tsolvers/lasolver/SparseMatrix.cc:29]: As discussed, let's keep it as it is for now.

**Review by blishko — APPROVED** [review]
LGTM!

**Pull request merged**